### DISCUSSION TITLE
Advice on Debugging Memory Leaks

### DISCUSSION THREAD

**Discussion by jpbro** [twinbasic/twinbasic#614] [Q&A]
I recently tried importing a VB6 app to TwinBasic ([described in detail here](https://www.vbforums.com/showthread.php?893990-VB6-Web-Site-App-Server&p=5550066&viewfull=1#post5550066)) and it went very well except for one lingering problem. There appears to be a memory leak in the tB compiled EXE that doesn't happen in the VB6 compiled one. I'm new to VS Code (and TwinBasic), so I just wanted to check in to see if any of you more experienced devs had an advice on how to best find the source of a memory leak in VS Code/tB?

Thanks in advance for any pointers.

**Comment by WaynePhillipsEA** [comment]
@jpbro Firstly, I have to say a big thank you for your feedback over at VBF.  There's some great feedback/ideas in that post that I plan to address.

The first thing I wanted to mention was that tB uses different memory allocation routines to VB6, and so it is not always directly comparable.  For example, you may find that the twinBASIC version does not actually release memory until the memory pressure grows as it is designed to reuse existing allocated memory in some instances.

That said, if the memory use is only going in one direction, then that probably is a leak.  My usual trick is to try to make the memory leak worse ... for example make string allocations much bigger, or perhaps add a big contiguous array to the end of a UDT.  The hope being that the leak starts to stick out like a sore thumb.   The next thing I would do is to add loops, for example if you had a function that generates a hash from a string, I might call it in a loop 10000 times, discarding the results except the last one, again with the aim of making the problem worse and easier to spot.

I see that you've got the source code over at VBF, so when I'm back in the office next week, I'll take a look at this for you if you like.

**Reply by jpbro** [reply]
Hi @WaynePhillipsEA , my pleasure regarding my feedback, I'm glad you found it useful.  A big thank you right back at you for your hard work on tB!

Thanks also for the info about tB's memory allocation, that' good information to have. It looks like memory is only going in one direction (well, it jumps a bit and drops a bit, but never below the previous jump so the trend is upwards).

I'm putting together a scaled down version of the app now, so I should hopefully have a better idea of where things are going wrong soon. Please don't waste your time looking any further into it right now - I'll let you know if I get stuck.

Happy New Year!

**Comment by jpbro** [comment]
I may have found the source of the problem in #618

**Reply by WaynePhillipsEA** [reply]
Thanks @jpbro, that one is now fixed.   I've also tweaked the twinBASIC panel layout, and the 'Debug Console' should now get the focus when the build button is pressed.  

There's still work to do on the twinBASIC panel layout, but it does feel better having the build button right next to the build configuration dropdown list now.
